File: sdks/tests/sign.cs
// 客户端 SDK 签名对齐测试脚本（C# 版）
// 用法：
//   dotnet script sign.csx -- <secret> <message>
// 或编译运行：
//   csc sign.cs /out:sign.exe
//   sign.exe <secret> <message>
// 输出：64 位小写 hex 签名（与后端 crypto.HMACSHA256 对齐）
//
// 优先尝试 BouncyCastle；不可用回退 HMACSHA256

using System;
using System.Reflection;
using System.Security.Cryptography;
using System.Text;

class Sign
{
    static int Main(string[] args)
    {
        if (args.Length != 2)
        {
            Console.Error.WriteLine("usage: sign.exe <secret> <message>");
            return 1;
        }
        Console.WriteLine(SignHmac(args[0], args[1]));
        return 0;
    }

    static string SignHmac(string secret, string msg)
    {
        byte[] key = Encoding.UTF8.GetBytes(secret);
        byte[] data = Encoding.UTF8.GetBytes(msg);

        // 探测 BouncyCastle
        var bcType = Type.GetType("Org.BouncyCastle.Crypto.Digests.Sha512_256Digest, BouncyCastle.Cryptography");
        if (bcType != null)
        {
            return BouncyCastleHex(key, data, bcType);
        }

        // 回退 HMACSHA256
        using var h = new HMACSHA256(key);
        byte[] raw = h.ComputeHash(data);
        return BitConverter.ToString(raw).Replace("-", "").ToLowerInvariant();
    }

    static string BouncyCastleHex(byte[] key, byte[] data, Type digestType)
    {
        var digest = Activator.CreateInstance(digestType)!;
        var macType = Type.GetType("Org.BouncyCastle.Crypto.Macs.HMac, BouncyCastle.Cryptography")!;
        var mac = Activator.CreateInstance(macType, digest)!;
        var keyParamType = Type.GetType("Org.BouncyCastle.Crypto.Parameters.KeyParameter, BouncyCastle.Cryptography")!;
        var keyParam = Activator.CreateInstance(keyParamType, key)!;
        macType.GetMethod("Init")!.Invoke(mac, new[] { keyParam });
        macType.GetMethod("BlockUpdate", new[] { typeof(byte[]), typeof(int), typeof(int) })!
            .Invoke(mac, new object[] { data, 0, data.Length });
        var result = new byte[32];
        macType.GetMethod("DoFinal", new[] { typeof(byte[]), typeof(int) })!
            .Invoke(mac, new object[] { result, 0 });
        return BitConverter.ToString(result).Replace("-", "").ToLowerInvariant();
    }
}

File: sdks/tests/Sign.java
// 客户端 SDK 签名对齐测试脚本（Java 版）
// 用法：java Sign <secret> <message>
// 输出：64 位小写 hex 签名（与后端 crypto.HMACSHA256 对齐）
//
// 编译：javac Sign.java
// 运行：java Sign <secret> <message>
//
// JDK 17+ 支持 HmacSHA512/256；JDK 11-16 回退 HmacSHA256
// 退出码 2 视为环境限制（JDK 不支持 HmacSHA512/256 且策略禁止回退）—— 当前实现总是回退，不退出 2

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;

public class Sign {

    public static void main(String[] args) throws Exception {
        if (args.length != 2) {
            System.err.println("usage: java Sign <secret> <message>");
            System.exit(1);
        }
        System.out.println(sign(args[0], args[1]));
    }

    static String sign(String secret, String msg) throws Exception {
        Mac mac;
        try {
            mac = Mac.getInstance("HmacSHA512/256");
        } catch (Exception e) {
            // 回退 HmacSHA256（与 SDK 一致）
            mac = Mac.getInstance("HmacSHA256");
        }
        mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), mac.getAlgorithm()));
        byte[] raw = mac.doFinal(msg.getBytes(StandardCharsets.UTF_8));
        return toHex(raw);
    }

    private static String toHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(String.format("%02x", b & 0xff));
        }
        return sb.toString();
    }
}
